tunnistamo/middleware.py

```diff
@@ -1,24 +1,43 @@
 import json
 import logging
 from datetime import timedelta
+from urllib.parse import parse_qsl, urlencode, urlsplit, urlunsplit
 
 from django.conf import settings
 from django.http import HttpResponseForbidden, HttpResponseRedirect
+from django.urls import reverse
 from django.utils import timezone
 from django.utils.http import quote
+from django.utils.translation import gettext_lazy as _
 from oidc_provider.lib.errors import BearerTokenError
+from social_core.exceptions import AuthException
 from social_django.middleware import SocialAuthExceptionMiddleware
 
 logger = logging.getLogger(__name__)
 
 
+def add_params_to_url(url, params):
+    """Add GET parameters to an existing url"""
+    url_parts = urlsplit(url)
+    new_query_parts = dict(parse_qsl(url_parts.query))
+    new_query_parts.update(params)
+    new_url_parts = list(url_parts)
+    new_url_parts[3] = urlencode(new_query_parts)
```

Comment:
Use named tuple syntax instead of hardcoded integer?

Reply:
The `url_parts` is a `SplitResult` which is immutable. The url_parts is converted to a list on the previous line.

I guess an alternative could be to create a new `SplitResult`. Something like this:

```python
return SplitResult(
    scheme=url_parts.scheme, 
    netloc=url_parts.netloc, 
    path=url_parts.path, 
    query=urlencode(new_query_parts),
    fragment=url_parts.fragment
).geturl()
```

Reply:
If `SplitResult` is a named tuple, as the documentation suggests, then it should have a [_replace](https://docs.python.org/3/library/collections.html#collections.somenamedtuple._replace) method.

Also, if you still have the `SplitResult` at hand, it should have a `geturl()` method, which returns the combined URL as string.

Reply:
Changed the `add_params_to_url` function.